README.md
# StyledText

Text widget with formatted text using tags. Makes it easier to use formatted text in multilingual applications.

Formatting is set in the text using **xml tags**, for which styles and other behaviors are defined separately. It is also possible to insert icons and widgets through tags.

You can set the click handler for the tag, through a tag definition class `StyledTextActionTag`.

**Attention!** The way of specifying the styles and behavior of tags has changed. See how to [migrate from version 2.0](#migration-from-version-20).

**Attention!** The default value of the `newLineAsBreaks` parameter has been changed, now it is enabled by default and line breaks are not ignored in the text.

## Table of Contents

- [Getting Started](#getting-Started)
  - [Escaping & special characters](#escaping--special-characters)
  - [Line breaks](#line-breaks)
- [Usage examples](#usage-examples)
  - [How to insert a widget into text](#an-example-of-inserting-an-input-field-widget-in-place-of-a-tag)
- [Migration from version 2.0](#migration-from-version-20)

## Getting Started

In your flutter project add the dependency:

```dart
dependencies:
  ...
  styled_text: ^[version]
```

Import package:
```dart
import 'package:styled_text/styled_text.dart';
```

### Escaping & special characters

Tag attributes must be enclosed in double quotes, for example: `<link href="https://flutter.dev">`.

You need to escape specific XML characters in text:
```
Original character  Escaped character
------------------  -----------------
"                   &quot;
'                   &apos;
&                   &amp;
<                   &lt;
>                   &gt;
<space>             &space;
```

### Line breaks

By default, line breaks are not ignored, all line breaks `\n` are automatically translated into the `<br/>` tag. To disable this behavior, you can set the `newLineAsBreaks` parameter to `false` and insert the `<br/>` tag where you want to break to a new line.

## Usage examples

### An example of making parts of text bold:
```dart
StyledText(
  text: 'Test: <bold>bold</bold> text.',
  tags: {
    'bold': StyledTextTag(style: TextStyle(fontWeight: FontWeight.bold)),
  },
)
```
![](https://github.com/andyduke/styled_text_package/raw/master/screenshots/1-bold.png)

---

### Example of highlighting a part of the text by different styles:
```dart
StyledText(
  text: 'Test: <bold>bold</bold> and <red>red color</red> text.',
  tags: {
    'bold': StyledTextTag(style: TextStyle(fontWeight: FontWeight.bold)),
    'red': StyledTextTag(style: TextStyle(fontWeight: FontWeight.bold, color: Colors.red)),
  },
)
```
![](https://github.com/andyduke/styled_text_package/raw/master/screenshots/2-bold-and-color.png)

---

### Example of inserting icons into the text:
```dart
StyledText(
  text: 'Text with alarm <alarm/> icon.',
  tags: {
    'alarm': StyledTextTagIcon(Icons.alarm),
  },
)
```
![](https://github.com/andyduke/styled_text_package/raw/master/screenshots/3-icon.png)

---

### Example of using a tag handler:
```dart
StyledText(
  text: 'Text with <link href="https://flutter.dev">link</link> inside.',
  tags: {
    'link': StyledTextTagAction(
      (TextSpan text, Map<String, String> attrs) => {
        final String link = attrs['href'];
        print('The "$link" link is tapped.');
      },
      style: TextStyle(decoration: TextDecoration.underline),
    ),
  },
)
```
![](https://github.com/andyduke/styled_text_package/raw/master/screenshots/4-link.png)

---

### Example of using a custom tag attributes handler, highlights text with the color specified in the "text" attribute of the tag:
```dart
StyledText(
  text: 'Text with custom <color text="#ff5500">color</color> text.',
  tags: {
    'color': StyledTextCustomTag(
        baseStyle: TextStyle(fontStyle: FontStyle.italic),
        parse: (baseStyle, attributes) {
          if (attributes.containsKey('text') &&
              (attributes['text'].substring(0, 1) == '#') &&
              attributes['text'].length >= 6) {
            final String hexColor = attributes['text'].substring(1);
            final String alphaChannel = (hexColor.length == 8) ? hexColor.substring(6, 8) : 'FF';
            final Color color = Color(int.parse('0x$alphaChannel' + hexColor.substring(0, 6)));
            return baseStyle.copyWith(color: color);
          } else {
            return baseStyle;
          }
        }),
  },
)
```
---

### An example of inserting an input field widget in place of a tag:
```dart
StyledText(
  text: 'Text with <input/> inside.',
  tags: {
    'input': StyledTextWidgetTag(
      TextField(
        decoration: InputDecoration(
          hintText: 'Input',
        ),
      ),
      size: Size.fromWidth(200),
      constraints: BoxConstraints.tight(Size(100, 50)),
    ),
  },
)
```
---

### An example of using a widget with the ability to select rich text:
```dart
StyledText.selectable(
  text: 'Test: <bold>bold</bold> text.',
  tags: {
    'bold': StyledTextTag(style: TextStyle(fontWeight: FontWeight.bold)),
  },
)
```

## Migration from version 2.0

Starting from version 3.0, the way of specifying styles and behavior for tags has changed. Now for this you need to use the `tags` parameter instead of `styles` and another set of classes to define styles and behavior.

Below are examples of converting old code to new.

### Specifying the text style

**OLD**
```dart
StyledText(
  text: 'Example: <b>bold</b> text.',
  styles: {
    'b': TextStyle(fontWeight: FontWeight.bold),
  },
)
```

**NEW**
```dart
StyledText(
  text: 'Example: <b>bold</b> text.',
  tags: {
    'b': StyledTextTag(style: TextStyle(fontWeight: FontWeight.bold)),
  },
)
```
---

### Specifying the icon

**OLD**
```dart
StyledText(
  text: 'Text with alarm <alarm/> icon.',
  styles: {
    'alarm': IconStyle(Icons.alarm),
  },
)
```

**NEW**
```dart
StyledText(
  text: 'Text with alarm <alarm/> icon.',
  tags: {
    'alarm': StyledTextIconTag(Icons.alarm),
  },
)
```
---

### Specifying a tap handler 

**OLD**
```dart
StyledText(
  text: 'Text with <link href="https://flutter.dev">link</link> inside.',
  styles: {
    'link': ActionTextStyle(
      decoration: TextDecoration.underline,
      onTap: (_, attrs) => _openLink(context, attrs),
    ),
  },
)
```

**NEW**
```dart
StyledText(
  text: 'Text with <link href="https://flutter.dev">link</link> inside.',
  tags: {
    'link': StyledTextActionTag(
      (_, attrs) => _openLink(context, attrs),
      style: TextStyle(decoration: TextDecoration.underline),
    ),
  },
)
```
---

### Specifying a custom parser for attributes and creating a style

**OLD**
```dart
StyledText(
  text: 'Text with custom <color text="#ff5500">color</color> text.',
  styles: {
    'color': CustomTextStyle(
        baseStyle: TextStyle(fontStyle: FontStyle.italic),
        parse: (baseStyle, attributes) {
          // Parser code here...
        }),
  },
)
```

**NEW**
```dart
StyledText(
  text: 'Text with custom <color text="#ff5500">color</color> text.',
  tags: {
    'color': StyledTextCustomTag(
        baseStyle: TextStyle(fontStyle: FontStyle.italic),
        parse: (baseStyle, attributes) {
          // Parser code here...
        }),
  },
)
```


CHANGELOG.md
## 3.0.2

* **[Breaking Changes]** In `StyledTextWidgetBuilderTag.builder` now the tag attributes are passed as the second parameter.

## 3.0.1

* Fixed a bug when unmounting a widget from a tree during text parsing.

## 3.0.0

* Flutter 2.2 compatibility.

## 3.0.0-beta+2

* **[Breaking Changes]** Changed the way text tags are styled. Now, instead of `TextStyle` and classes that implement it, you need to use new classes that describe the styles and behavior of tags.
* **[Breaking changes]** The `newLineAsBreaks` parameter is now `true` by default.
* Added the ability to insert any widgets into the text.

## 2.0.0

* Fixed a typo in the example.

## 2.0.0-nullsafety.0

* Migrate to null safety.

## 1.0.3+4

* Improved handling of broken xml.

## 1.0.3+3

* Fixed a bug when the space between two tags was eaten.

## 1.0.3+2

* Made a workaround to keep whitespace between tags: `<b>bold</b>&space;<i>italic</i>`.
* Added `size`, `color` and `backgroundColor` parameters to `IconStyle` class.

## 1.0.3+1

* Added `CustomTextStyle` text style, for which you can specify handling of tag attributes.

## 1.0.2+1

* Added the `StyledText.selectable` constructor to create selectable text.
* Added guidance on escaping XML special characters in text.
* The deprecated `isNewLineAsBreaks` parameter has been removed.

## 1.0.1+3

* Fixed flickering when changing style, styles and text.

## 1.0.1+2

* Improved support for DefaultTextSyle, now `style` property is merging with DefaultTextSyle.
* Fixed a bug where no redrawing occurred when changing `style` property.

## 1.0.1+1

* The parameter isNewLineAsBreaks has been renamed to newLineAsBreaks, isNewLineAsBreaks is deprecated.

## 1.0.1

* The parameter isNewLineAsBreaks has been added, indicating not to ignore line breaks in the source text.
* Added description of parameters via doc comments.

## 1.0.0+2

* Allow XML special chars: &lt; (<), &amp; (&), &gt; (>), &quot; ("), and &apos; (')
* Avoid memory leak log message

## 1.0.0+1

* Example added.

## 1.0.0

* Initial release.


example/lib/main.dart
import 'package:flutter/material.dart';
import 'package:styled_text/styled_text.dart';

void main() => runApp(MyApp());

class MyApp extends StatelessWidget {
  @override
  Widget build(BuildContext context) {
    return MaterialApp(
      title: 'StyledText Demo',
      theme: ThemeData(
        primarySwatch: Colors.teal,
      ),
      home: DemoPage(),
    );
  }
}

class DemoPage extends StatelessWidget {
  void _alert(BuildContext context, {String text = 'Tapped'}) {
    showDialog<void>(
      context: context,
      builder: (BuildContext context) {
        return AlertDialog(
          content: Text(text),
          actions: <Widget>[
            TextButton(
              child: Text('OK'),
              onPressed: () {
                Navigator.of(context).pop();
              },
            ),
          ],
        );
      },
    );
  }

  void _openLink(BuildContext context, Map<String?, String?> attrs) {
    final String? link = attrs['href'];

    showDialog<void>(
      context: context,
      builder: (BuildContext context) {
        return AlertDialog(
          title: Text('Open Link'),
          content: Text(link ?? 'Unknown link'),
          actions: <Widget>[
            TextButton(
              child: Text('Ok'),
              onPressed: () {
                Navigator.of(context).pop();
              },
            ),
          ],
        );
      },
    );
  }

  @override
  Widget build(BuildContext context) {
    return Scaffold(
      appBar: AppBar(
        title: Text('StyledText Demo'),
      ),
      body: Center(
        child: SingleChildScrollView(
          padding: const EdgeInsets.symmetric(vertical: 32),
          child: Column(
            mainAxisAlignment: MainAxisAlignment.center,
            children: <Widget>[
              // Simple formatted text
              StyledText(
                text: 'Test: <b>bold</b> text.',
                tags: {
                  'b': StyledTextTag(
                      style: TextStyle(fontWeight: FontWeight.bold)),
                },
              ),

              // Nested multiple styles
              StyledText(
                text: 'Test: <b>bold <i>italic</i> bold</b> text.',
                tags: {
                  'b': StyledTextTag(
                      style: TextStyle(fontWeight: FontWeight.bold)),
                  'i': StyledTextTag(
                      style: TextStyle(fontStyle: FontStyle.italic)),
                },
              ),

              // Text with quotes
              StyledText(
                text: 'Quote test: <b>&quot;bold&quot;</b> text.',
                tags: {
                  'b': StyledTextTag(
                      style: TextStyle(fontWeight: FontWeight.bold)),
                },
              ),

              // Multiline text without breaks
              const SizedBox(height: 20),
              StyledText(
                newLineAsBreaks: false,
                text: """Multiline text 
(wo breaks)""",
                tags: {
                  'b': StyledTextTag(
                      style: TextStyle(fontWeight: FontWeight.bold)),
                },
              ),

              // Multiline text with breaks
              const SizedBox(height: 20),
              StyledText(
                text: """Multiline text
(with breaks)""",
                tags: {
                  'b': StyledTextTag(
                      style: TextStyle(fontWeight: FontWeight.bold)),
                },
              ),

              // Text with icon
              const SizedBox(height: 20),
              StyledText(
                text: 'Text with alarm <alarm/> icon.',
                tags: {
                  'alarm': StyledTextIconTag(
                    Icons.alarm,
                    color: Colors.teal,
                    size: 18,
                    onTap: (text, attributes) =>
                        _alert(context, text: 'Alarm Tapped'),
                  ),
                },
              ),

              // Text with icon inside styled text
              const SizedBox(height: 20),
              StyledText(
                text: 'Text with <red>alarm <alarm/> icon</red>.',
                tags: {
                  'red': StyledTextTag(style: TextStyle(color: Colors.red)),
                  'alarm': StyledTextIconTag(
                    Icons.alarm,
                    color: Colors.teal,
                    size: 18,
                    onTap: (text, attributes) =>
                        _alert(context, text: 'Alarm Tapped'),
                  ),
                },
              ),

              // Text with link
              const SizedBox(height: 20),
              StyledText(
                text:
                    'Text with <link href="https://flutter.dev">link</link> inside.',
                tags: {
                  'link': StyledTextActionTag(
                    (_, attrs) => _openLink(context, attrs),
                    style: TextStyle(decoration: TextDecoration.underline),
                  ),
                },
              ),

              // Text with action
              const SizedBox(height: 20),
              StyledText(
                text:
                    'Text with <action><red>red</red> action</action> inside.',
                tags: {
                  'red': StyledTextTag(style: TextStyle(color: Colors.red)),
                  'action': StyledTextActionTag(
                    (text, attributes) => _alert(context),
                    style: TextStyle(decoration: TextDecoration.underline),
                  ),
                },
              ),

              // Text with widget
              const SizedBox(height: 20),
              StyledText(
                text: 'Text with <input/> inside.',
                tags: {
                  'input': StyledTextWidgetTag(
                    TextField(
                      decoration: InputDecoration(
                        hintText: 'Input',
                      ),
                    ),
                    size: Size.fromWidth(200),
                    constraints: BoxConstraints.tight(Size(100, 50)),
                  ),
                },
              ),

              const Divider(height: 40),

              // Selectable text
              StyledText.selectable(
                text: 'Test: selectable <b>bold</b> text.',
                tags: {
                  'b': StyledTextTag(
                      style: TextStyle(fontWeight: FontWeight.bold)),
                },
              ),

              const Divider(height: 40),

              // Text with custom color tag
              StyledText(
                text:
                    'Text with custom <color text="#ff5500">color</color> text.',
                tags: {
                  'color': StyledTextCustomTag(
                      baseStyle: TextStyle(fontStyle: FontStyle.italic),
                      parse: (baseStyle, attributes) {
                        if (attributes.containsKey('text') &&
                            (attributes['text']!.substring(0, 1) == '#') &&
                            attributes['text']!.length >= 6) {
                          final String hexColor =
                              attributes['text']!.substring(1);
                          final String alphaChannel = (hexColor.length == 8)
                              ? hexColor.substring(6, 8)
                              : 'FF';
                          final Color color = Color(int.parse(
                              '0x$alphaChannel' + hexColor.substring(0, 6)));
                          return baseStyle?.copyWith(color: color);
                        } else {
                          return baseStyle;
                        }
                      }),
                },
              ),
            ],
          ),
        ),
      ),
    );
  }
}
